File: eclipse-plugin/src/main/java/xyz/devglobe/eclipse/core/DevGlobeTracker.java
package xyz.devglobe.eclipse.core;

import xyz.devglobe.eclipse.auth.ConfigWriter;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Central tracker orchestrating the core client lifecycle, document listening,
 * and state management. Mirrors DevGlobeTracker from the JetBrains plugin.
 */
public class DevGlobeTracker {

    private static final String PLUGIN_VERSION = "2.0.1";
    private static final String EDITOR = "eclipse";

    // ── Singleton ────────────────────────────────────────────────────────

    private static final DevGlobeTracker INSTANCE = new DevGlobeTracker();

    public static DevGlobeTracker getInstance() {
        return INSTANCE;
    }

    // ── State ────────────────────────────────────────────────────────────

    private volatile TrackerState state = TrackerState.DEFAULT;
    private final List<Runnable> stateListeners = new ArrayList<>();
    private final AtomicBoolean starting = new AtomicBoolean(false);
    private final AtomicBoolean intentionalShutdown = new AtomicBoolean(false);
    private CoreClient client;

    // ── Public API ───────────────────────────────────────────────────────

    public TrackerState getState() {
        return state;
    }

    public void addStateListener(Runnable listener) {
        synchronized (stateListeners) {
            stateListeners.add(listener);
        }
    }

    public void removeStateListener(Runnable listener) {
        synchronized (stateListeners) {
            stateListeners.remove(listener);
        }
    }

    private void fireStateChanged() {
        List<Runnable> listeners;
        synchronized (stateListeners) {
            listeners = new ArrayList<>(stateListeners);
        }
        for (Runnable l : listeners) {
            try { l.run(); } catch (Exception ignored) {}
        }
    }

    // ── Lifecycle ────────────────────────────────────────────────────────

    public void saveApiKeyAndStart(String apiKey) {
        ConfigWriter.writeApiKey(apiKey);
        start();
    }

    public void start() {
        if (!starting.compareAndSet(false, true)) return;
        if (!ConfigWriter.hasApiKey()) {
            updateState(state.withConfigured(false).withError("No API key configured"));
            starting.set(false);
            return;
        }
        ensureCore();
    }

    public void pause() {
        if (client != null && client.isRunning()) {
            client.sendPause();
            updateState(state.withTracking(false));
            Notifier.confirm("Tracking paused");
        }
    }

    public void resume() {
        if (client != null && client.isRunning()) {
            client.sendResume();
            updateState(state.withTracking(true));
            Notifier.confirm("Tracking started");
        }
    }

    public void reset() {
        shutdownCore();
        ConfigWriter.clearApiKey();
        updateState(TrackerState.DEFAULT);
    }

    public void sendSetStatus(String message) {
        if (client != null && client.isRunning()) {
            client.sendSetStatus(message);
        }
    }

    public void shutdown() {
        shutdownCore();
        updateState(TrackerState.DEFAULT);
    }

    // ── Activity ────────────────────────────────────────────────────────

    public void sendActivity(String filePath, String language) {
        if (client != null && client.isRunning() && state.tracking) {
            client.sendActivity(filePath, language);
            // Update local language immediately so the UI reflects it
            // without waiting for the next heartbeat_ok from core
            if (language != null && !language.isEmpty() && !language.equals(state.language)) {
                updateState(state.withLanguage(language));
            }
        }
    }

    // ── Core management ─────────────────────────────────────────────────

    private void ensureCore() {
        intentionalShutdown.set(false);
        if (client != null && client.isRunning()) {
            // Already running — re-init
            client.sendInit(PLUGIN_VERSION, EDITOR);
            return;
        }

        if (!CoreDownloader.isInstalled()) {
            DevGlobePlugin.log("devglobe-core not found, downloading...");
            updateState(state.withError("Downloading devglobe-core..."));
            boolean ok = CoreDownloader.download();
            if (!ok) {
                updateState(state.withError("Failed to download devglobe-core"));
                starting.set(false);
                return;
            }
        }

        File binary = CoreDownloader.getBinaryPath();
        client = new CoreClient(new CoreCallbacks());
        boolean ok = client.start(binary.getAbsolutePath());
        if (!ok) {
            updateState(state.withError("Failed to start devglobe-core"));
            starting.set(false);
            return;
        }
        client.sendInit(PLUGIN_VERSION, EDITOR);
    }

    private void shutdownCore() {
        intentionalShutdown.set(true);
        if (client != null) {
            client.stop();
            client = null;
        }
        // Keep intentionalShutdown=true until onProcessDied() has run.
        // It will be reset at the start of the next ensureCore() call.
    }

    // ── State updates ────────────────────────────────────────────────────

    private void updateState(TrackerState newState) {
        this.state = newState;
        fireStateChanged();
    }

    // ── Core callbacks ──────────────────────────────────────────────────

    private class CoreCallbacks implements CoreClient.Callbacks {
        @Override
        public void onReady(boolean configured) {
            starting.set(false);
            if (configured) {
                // Don't assume online — wait for the actual "online" event from core.
                // Set tracking=true so the UI shows the dashboard, but keep offline=true
                // until the core confirms connectivity via onOnline().
                updateState(state.withConfigured(true).withTracking(true).withError(null).withOffline(true));
            } else {
                updateState(state.withConfigured(false).withError("Not configured"));
            }
        }

        @Override
        public void onNotConfigured() {
            starting.set(false);
            updateState(state.withConfigured(false).withError("Not configured — check your API key"));
        }

        @Override
        public void onInvalidApiKey() {
            starting.set(false);
            updateState(state.withConfigured(false).withError("Invalid API key"));
            Notifier.error("Invalid API key");
        }

        @Override
        public void onHeartbeatOk(int todaySeconds, String language) {
            String ct = TrackerState.formatSeconds(todaySeconds);
            TrackerState updated = state.withTodaySeconds(todaySeconds)
                    .withCodingTime(ct)
                    .withTracking(true)
                    .withOffline(false)
                    .withError(null);
            if (language != null && !language.isEmpty()) {
                updated = updated.withLanguage(language);
            }
            updateState(updated);
        }

        @Override
        public void onOffline() {
            updateState(state.withOffline(true));
        }

        @Override
        public void onOnline() {
            updateState(state.withOffline(false).withError(null));
            Notifier.confirm("Connected!");
        }

        @Override
        public void onStatusOk() {
            DevGlobePlugin.log("Status set successfully");
            Notifier.confirm("Status updated");
        }

        @Override
        public void onStatusError(String message) {
            DevGlobePlugin.log("Status error: " + message);
            updateState(state.withError(message));
            Notifier.error("Status error: " + message);
        }

        @Override
        public void onError(String message) {
            starting.set(false);
            DevGlobePlugin.log("Core error: " + message);
            updateState(state.withError(message));
            Notifier.error(message);
        }

        @Override
        public void onProcessDied() {
            starting.set(false);
            if (intentionalShutdown.get()) {
                // Expected during disconnect/reset - don't show error
                DevGlobePlugin.log("Core process stopped intentionally");
            } else {
                // Unexpected process death
                DevGlobePlugin.log("Core process died unexpectedly");
                updateState(state.withError("Core process died").withTracking(false));
                Notifier.error("Core process died");
            }
        }
    }
}

File: jetbrains-plugin/src/main/kotlin/xyz/devglobe/plugin/core/DevGlobeTracker.kt
package xyz.devglobe.plugin.core

import com.intellij.notification.Notification
import com.intellij.notification.NotificationType
import com.intellij.openapi.Disposable
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.editor.event.DocumentEvent
import com.intellij.openapi.editor.event.DocumentListener
import com.intellij.openapi.editor.EditorFactory
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.ide.plugins.PluginManagerCore
import com.intellij.openapi.extensions.PluginId
import xyz.devglobe.plugin.auth.ApiKeyStorage
import xyz.devglobe.plugin.auth.ConfigWriter
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.atomic.AtomicBoolean

class DevGlobeTracker : Disposable {

    private val LOG = Logger.getInstance(DevGlobeTracker::class.java)

    @Volatile private var state = TrackerState()
    @Volatile private var coreClient: CoreClient? = null
    private val started = AtomicBoolean(false)
    private var documentListener: DocumentListener? = null
    private val stateListeners = CopyOnWriteArrayList<(TrackerState) -> Unit>()

    fun getState(): TrackerState = state.copy()

    fun addStateListener(listener: (TrackerState) -> Unit) {
        stateListeners.add(listener)
    }

    fun removeStateListener(listener: (TrackerState) -> Unit) {
        stateListeners.remove(listener)
    }

    /** Called when the user enters their API key. Writes to config.toml + OS keychain, then starts. */
    fun saveApiKeyAndStart(apiKey: String) {
        ConfigWriter.writeApiKey(apiKey)
        ApiKeyStorage.set(apiKey)
        start()
    }

    fun start() {
        if (!started.compareAndSet(false, true)) return

        updateState { it.copy(tracking = true) }

        ApplicationManager.getApplication().executeOnPooledThread {
            if (!ensureCore()) {
                started.set(false)
                updateState { it.copy(tracking = false) }
                return@executeOnPooledThread
            }
            coreClient?.sendInit(pluginVersion())
            coreClient?.sendResume()
            ApplicationManager.getApplication().invokeLater { registerDocumentListener() }
        }
    }

    fun pause() {
        started.set(false)
        coreClient?.sendPause()
        updateState { it.copy(tracking = false) }
    }

    fun reset() {
        started.set(false)
        coreClient?.sendShutdown()
        coreClient?.dispose()
        synchronized(this) { coreClient = null }
        ConfigWriter.clearApiKey()
        ApiKeyStorage.clear()
        updateState { TrackerState() }
    }

    fun restoreConnected() {
        updateState { it.copy(configured = true, tracking = false, error = null) }
    }

    fun sendSetStatus(message: String) {
        ApplicationManager.getApplication().executeOnPooledThread {
            ensureCore()
            coreClient?.sendSetStatus(message)
        }
    }

    @Synchronized
    private fun ensureCore(): Boolean {
        if (coreClient?.isAlive() == true) return true

        coreClient?.dispose()
        coreClient = null

        if (!CoreDownloader.isInstalled() && !downloadCore()) return false

        val client = CoreClient(CoreDownloader.getBinaryPath()).also(::wireCallbacks)
        if (!client.start()) {
            updateState { it.copy(error = "Failed to start devglobe-core") }
            return false
        }
        coreClient = client
        return true
    }

    private fun downloadCore(): Boolean {
        LOG.info("devglobe-core not found, downloading...")
        notify("Downloading devglobe-core...", NotificationType.INFORMATION)
        if (!CoreDownloader.download()) {
            CoreDownloader.notifyDownloadFailed()
            updateState { it.copy(error = "Failed to download devglobe-core") }
            return false
        }
        notify("devglobe-core downloaded successfully", NotificationType.INFORMATION)
        return true
    }

    private fun wireCallbacks(client: CoreClient) {
        client.onReady = { configured ->
            updateState { it.copy(configured = configured, error = null) }
        }
        client.onNotConfigured = {
            updateState { it.copy(configured = false, tracking = false) }
        }
        client.onInvalidApiKey = {
            LOG.warn("API key rejected by server (401), clearing local key")
            ApiKeyStorage.clear()
            ConfigWriter.clearApiKey()
            updateState { it.copy(configured = false, tracking = false) }
            notify(
                "Invalid API key. Please reconnect with a valid key from devglobe.app/dashboard/settings.",
                NotificationType.ERROR,
            )
        }
        client.onHeartbeatOk = { todaySeconds, language ->
            updateState {
                it.copy(
                    todaySeconds = todaySeconds,
                    codingTime = formatSeconds(todaySeconds),
                    language = language,
                    tracking = true,
                    offline = false,
                )
            }
        }
        client.onOffline = { updateState { it.copy(offline = true) } }
        client.onOnline = { updateState { it.copy(offline = false) } }
        client.onStatusOk = { notify("Status updated", NotificationType.INFORMATION) }
        client.onStatusError = { msg -> notify(msg, NotificationType.ERROR) }
        client.onError = { msg ->
            notify(msg, NotificationType.ERROR)
            updateState { it.copy(error = msg) }
        }
        client.onProcessDied = {
            LOG.warn("devglobe-core process died unexpectedly")
            synchronized(this) {
                coreClient = null
                started.set(false)
            }
            updateState { it.copy(tracking = false, error = "devglobe-core stopped unexpectedly") }
            notify("DevGlobe tracking stopped — devglobe-core crashed", NotificationType.WARNING)
        }
    }

    private fun registerDocumentListener() {
        if (documentListener != null) return
        val listener = object : DocumentListener {
            override fun documentChanged(event: DocumentEvent) {
                val project = LanguageService.getFocusedProject() ?: return
                val editor = FileEditorManager.getInstance(project).selectedTextEditor ?: return
                val file = editor.virtualFile ?: return
                coreClient?.sendActivity(file.path, LanguageService.detectLanguage(file))
            }
        }
        documentListener = listener
        EditorFactory.getInstance().eventMulticaster.addDocumentListener(listener, this)
    }

    private fun updateState(transform: (TrackerState) -> TrackerState) {
        val snapshot = synchronized(this) {
            state = transform(state)
            state.copy()
        }
        ApplicationManager.getApplication().invokeLater {
            for (listener in stateListeners) {
                listener(snapshot)
            }
        }
    }

    private fun notify(message: String, type: NotificationType) {
        ApplicationManager.getApplication().invokeLater {
            Notification("DevGlobe", "DevGlobe", message, type).notify(null)
        }
    }

    private fun pluginVersion(): String {
        return try {
            PluginManagerCore.getPlugin(PluginId.getId("xyz.devglobe.plugin"))?.version ?: "0.0.0"
        } catch (_: Exception) {
            "0.0.0"
        }
    }

    private fun formatSeconds(seconds: Int): String {
        val h = seconds / 3600
        val m = (seconds % 3600) / 60
        return if (h > 0) "${h}h ${m}m" else "${m}m"
    }

    override fun dispose() {
        coreClient?.dispose()
        coreClient = null
    }

    companion object {
        fun getInstance(): DevGlobeTracker =
            ApplicationManager.getApplication().getService(DevGlobeTracker::class.java)
    }
}
